File: src/message.rs
use serde::{Serialize, Deserialize};

pub enum ServerStates {
  WaitingForConnection,
  ExchangingKeys,
  AuthenticatingClient,
  WaitingForMessages,
}

pub enum ClientStates {
  StartingConnection,
  ExchangingKeys,
  Authenticating,
  SendingMessage,
}

#[derive(Serialize, Deserialize, Debug)]
pub enum ClientMessages {
  SimKey { key_bytes : Vec<u8>, iv_bytes : Vec<u8> },
  Auth { hashed_password_bytes : Vec<u8>, upd_port_listening : u16 },
  MovePointer { x : i32, y : i32 },
  PressKey { key_codes : [u8; 32]},
  RunCommand { current : u8, total : u8, string_bytes : Vec<u8> },
  Quit,
}

#[derive(Serialize, Deserialize, Debug)]
pub enum ServerMessages {
  PubKey { der_bytes : Vec<u8> },
  AccessGranted { upd_port_listening : u16 },
  AccessDenied,
  Ack { sequence_aknowledgement : u32 },
  RequestSend { sequence_requested : u32 },
}

#[derive(Serialize, Deserialize, Debug)]
pub enum MessageType {
  Client(ClientMessages),
  Server(ServerMessages)
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Message {
  pub sequence : u32,
  pub message_type : MessageType,
}

impl Message {
  pub fn new(sequence: u32, message_type : MessageType) -> Self {
    Self {
      sequence : sequence,
      message_type : message_type,
    }
  }

  pub fn ser(&self) -> Vec<u8> {
    serde_cbor::to_vec(&self).unwrap()
  }

  pub fn des(slice: &[u8]) -> Option<Self> {
    let res = serde_cbor::from_slice(slice);
    match res {
      Ok(message) => Some(message),
      Err(err) =>{
        println!("Eror {:?}", err);
        None 
      },
    }
  }
}

File: src/python/message.py
from enum import Enum
class ClientStates(Enum):
  StartingConnection = 0,
  ExchangingKeys = 1,
  Authenticating = 2,
  SendingMessage = 3,

message_sequence = 0

def creteSimKeyMessage(key_bytes:bytes, iv_bytes:bytes) -> dict:
  global message_sequence
  message_sequence += 1
  return {
    'sequence' : message_sequence,
    'message_type' : {
      'Client' : {
        'SimKey' : {
          'key_bytes' : [int(x) for x in key_bytes],
          'iv_bytes' : [int(x) for x in iv_bytes]
        }
      }
    }
  }

def creteAuthMessage(hashed_password_bytes:bytes, upd_port_listening:int) -> dict:
  global message_sequence
  message_sequence += 1
  return {
    'sequence' : message_sequence,
    'message_type' : {
      'Client' : {
        'Auth' : {
          'hashed_password_bytes' : [int(x) for x in hashed_password_bytes],
          'upd_port_listening' : upd_port_listening
        }
      }
    }
  }

def creteMovePointerMessage(x:int, y:int) -> dict:
  global message_sequence
  message_sequence += 1
  return {
    'sequence' : message_sequence,
    'message_type' : {
      'Client' : {
        'MovePointer' : {
          'x' : x,
          'y' : y
        }
      }
    }
  }

def cretePressKeyMessage(key_codes : [int]) -> dict:
  global message_sequence
  message_sequence += 1
  return {
    'sequence' : message_sequence,
    'message_type' : {
      'Client' : {
        'PressKey' : {
          'key_codes' : key_codes[0:32]
        }
      }
    }
  }

def creteRunCommandMessage(current : int, total : int, string_bytes : bytes ) -> dict:
  global message_sequence
  message_sequence += 1
  return {
    'sequence' : message_sequence,
    'message_type' : {
      'Client' : {
        'RunCommand' : {
          'current' : current,
          'total' : total,
          'string_bytes' : [int(x) for x in string_bytes]
        }
      }
    }
  }

def creteQuitMessage() -> dict:
  global message_sequence
  message_sequence += 1
  return {
    'sequence' : message_sequence,
    'message_type' : {
      'Client' : 'Quit',
    }
  }